server/mqtt.go

```diff
@@ -3440,15 +3793,77 @@ func (c *client) mqttProcessPubAck(pi uint16) {
 		if len(sess.pending) == 0 {
 			sess.ppi = 0
 		}
-		ackSubject = ack.subject
+		jsAckSubject = ack.jsAckSubject
 	}
-	// Send the ack if applicable.
-	if ackSubject != _EMPTY_ {
```

Comment:
@kozlovic why were lines 3446-3449 inside the lock? As I understand it, sending the ack is async anyway, and the send queue has its own synchronization?

Reply:
True, but the "push" should be fast and non blocking also, so you would not gain too much by doing the processing outside, and you will introduce more changes in this PR than needed. You could have a later pass at code to try to reduce time under lock, but I would keep changes in this PR to a strict minimum to implement QoS2.
All that being said, yes, sess.jsa is immutable (so is the sess.jsa.sendq) and therefore you could that outside of the lock - given that this function will be invoked by the readLoop routine...